File: implementations/go/pocketplus/compressor.go
package pocketplus

import (
	"errors"
	"fmt"
)

// Constants
const (
	MaxHistory    = 16 // History depth for change vectors
	MaxVtHistory  = 16 // History size for Vt calculation
	MaxRobustness = 7  // Maximum robustness level
)

// CompressParams holds compression parameters for a single packet.
type CompressParams struct {
	MinRobustness    int  // Rt: Minimum robustness level (0-7)
	NewMaskFlag      bool // pt: Update mask from build vector
	SendMaskFlag     bool // ft: Include mask in output
	UncompressedFlag bool // rt: Send uncompressed
}

// Compressor maintains state for POCKET+ compression.
type Compressor struct {
	// Configuration (immutable after init)
	F          int // Input vector length in bits
	robustness int // Rt: Base robustness level (0-7)

	// Period limits for automatic parameter management
	ptLimit int
	ftLimit int
	rtLimit int

	// State (updated each cycle)
	mask        *BitVector
	prevMask    *BitVector
	build       *BitVector
	prevInput   *BitVector
	initialMask *BitVector

	// Change history (circular buffer)
	changeHistory [MaxHistory]*BitVector
	historyIndex  int

	// Flag history for ct calculation
	newMaskFlagHistory [MaxVtHistory]int
	flagHistoryIndex   int

	// Cycle counter
	t int

	// Countdown counters for automatic mode
	ptCounter int
	ftCounter int
	rtCounter int

	// Pre-allocated working buffers (avoid per-packet allocations)
	workChange      *BitVector // For change computation
	workXt          *BitVector // For robustness window
	workCombined    *BitVector // For combining changes
	workInvMask     *BitVector // For inverted mask
	workExtractMask *BitVector // For extraction mask
	workMaskShifted *BitVector // For mask shift
	workMaskDiff    *BitVector // For mask XOR
	workChanges     *BitVector // For input XOR prevInput
	workOutput      *BitBuffer // For output buffer
}

// NewCompressor creates a new compressor.
func NewCompressor(F int, initialMask *BitVector, robustness, ptLimit, ftLimit, rtLimit int) (*Compressor, error) {
	if F <= 0 {
		return nil, errors.New("F must be positive")
	}
	if robustness < 0 || robustness > MaxRobustness {
		return nil, fmt.Errorf("robustness must be between 0 and %d", MaxRobustness)
	}

	comp := &Compressor{
		F:          F,
		robustness: robustness,
		ptLimit:    ptLimit,
		ftLimit:    ftLimit,
		rtLimit:    rtLimit,
	}

	// Initialize bit vectors
	var err error
	comp.mask, err = NewBitVector(F)
	if err != nil {
		return nil, err
	}
	comp.prevMask, _ = NewBitVector(F)
	comp.build, _ = NewBitVector(F)
	comp.prevInput, _ = NewBitVector(F)
	comp.initialMask, _ = NewBitVector(F)

	// Initialize change history
	for i := 0; i < MaxHistory; i++ {
		comp.changeHistory[i], _ = NewBitVector(F)
	}

	// Initialize working buffers
	comp.workChange, _ = NewBitVector(F)
	comp.workXt, _ = NewBitVector(F)
	comp.workCombined, _ = NewBitVector(F)
	comp.workInvMask, _ = NewBitVector(F)
	comp.workExtractMask, _ = NewBitVector(F)
	comp.workMaskShifted, _ = NewBitVector(F)
	comp.workMaskDiff, _ = NewBitVector(F)
	comp.workChanges, _ = NewBitVector(F)
	comp.workOutput = NewBitBuffer()

	// Set initial mask if provided
	if initialMask != nil {
		comp.initialMask.CopyFrom(initialMask)
		comp.mask.CopyFrom(initialMask)
	}

	// Reset state
	comp.Reset()

	return comp, nil
}

// Reset resets the compressor to initial state.
func (comp *Compressor) Reset() {
	comp.t = 0
	comp.historyIndex = 0

	// Reset mask to initial
	comp.mask.CopyFrom(comp.initialMask)
	comp.prevMask.Zero()

	// Clear build and prevInput
	comp.build.Zero()
	comp.prevInput.Zero()

	// Clear change history
	for i := 0; i < MaxHistory; i++ {
		comp.changeHistory[i].Zero()
	}

	// Clear flag history
	for i := 0; i < MaxVtHistory; i++ {
		comp.newMaskFlagHistory[i] = 0
	}
	comp.flagHistoryIndex = 0

	// Reset countdown counters
	comp.ptCounter = comp.ptLimit
	comp.ftCounter = comp.ftLimit
	comp.rtCounter = comp.rtLimit
}

// CompressPacket compresses a single input packet.
func (comp *Compressor) CompressPacket(input *BitVector, params *CompressParams) ([]byte, error) {
	if input == nil || input.length != comp.F {
		return nil, errors.New("input must be non-nil and match F length")
	}

	// Use default params if none provided
	if params == nil {
		params = &CompressParams{MinRobustness: comp.robustness}
	}

	// Reuse pre-allocated output buffer
	comp.workOutput.Clear()
	output := comp.workOutput

	// ================================================================
	// STEP 1: Update Mask and Build Vectors (CCSDS Section 4)
	// ================================================================

	// Save previous mask (reuse prevMask field)
	comp.prevMask.CopyFrom(comp.mask)
	prevMask := comp.prevMask

	// Save previous build into workCombined temporarily
	comp.workCombined.CopyFrom(comp.build)
	prevBuild := comp.workCombined

	// Update build vector (Equation 6)
	if comp.t > 0 {
		updateBuildInternal(comp.build, input, comp.prevInput, comp.workChanges, params.NewMaskFlag, comp.t)
	}

	// Update mask vector (Equation 7)
	if comp.t > 0 {
		updateMaskInternal(comp.mask, input, comp.prevInput, prevBuild, comp.workChanges, params.NewMaskFlag)
	}

	// Compute change vector (Equation 8) - reuse workChange
	change := comp.workChange
	computeChangeInternal(change, comp.mask, prevMask, comp.t)

	// Store change in history (circular buffer)
	comp.changeHistory[comp.historyIndex].CopyFrom(change)

	// ================================================================
	// STEP 2: Encode Output Packet (CCSDS Section 5.3)
	// ot = ht || qt || ut
	// ================================================================

	// Calculate Xt (robustness window) - reuse workXt
	Xt := comp.computeRobustnessWindowInto(change, comp.workXt)

	// Calculate Vt (effective robustness)
	Vt := comp.computeEffectiveRobustness(change)

	// Calculate dt flag
	var dt int
	if !params.SendMaskFlag && !params.UncompressedFlag {
		dt = 1
	}

	// ================================================================
	// Component ht: Mask change information
	// ht = RLE(Xt) || BIT4(Vt) || et || kt || ct || dt
	// ================================================================

	// 1. RLE(Xt) - Run-length encode the robustness window
	RLEEncode(output, Xt)

	// 2. BIT4(Vt) - 4-bit effective robustness level
	output.AppendValue(uint64(Vt), 4)

	// 3. et, kt, ct - Only if Vt > 0 and there are mask changes
	if Vt > 0 && Xt.HammingWeight() > 0 {
		// Calculate et
		et := hasPositiveUpdates(Xt, comp.mask)
		output.AppendBit(et)

		if et != 0 {
			// kt - Output '1' for positive updates (mask=0), '0' for negative
			// Extract INVERTED mask values in forward order - reuse workInvMask
			invertedMask := comp.workInvMask
			for j := 0; j < comp.mask.length; j++ {
				maskBit := comp.mask.GetBit(j)
				if maskBit == 0 {
					invertedMask.SetBit(j, 1)
				} else {
					invertedMask.SetBit(j, 0)
				}
			}
			BitExtractForward(output, invertedMask, Xt)

			// Calculate and encode ct
			ct := comp.computeCtFlag(Vt, params.NewMaskFlag)
			output.AppendBit(ct)
		}
	}

	// 4. dt - Flag indicating if both ft and rt are zero
	output.AppendBit(dt)

	// ================================================================
	// Component qt: Optional full mask
	// qt = empty if dt=1, '1' || RLE(<(Mt XOR (Mt<<))>) if ft=1, '0' otherwise
	// ================================================================

	if dt == 0 { // Only if dt = 0
		if params.SendMaskFlag {
			output.AppendBit(1) // Flag: mask follows

			// Encode mask as RLE(M XOR (M<<)) - reuse working buffers
			leftShiftInto(comp.workMaskShifted, comp.mask)
			comp.workMaskDiff.XORInto(comp.mask, comp.workMaskShifted)
			RLEEncode(output, comp.workMaskDiff)
		} else {
			output.AppendBit(0) // Flag: no mask
		}
	}

	// ================================================================
	// Component ut: Unpredictable bits or full input
	// ================================================================

	if params.UncompressedFlag {
		// '1' || COUNT(F) || It
		output.AppendBit(1) // Flag: full input follows
		CountEncode(output, comp.F)
		output.AppendBitVector(input)
	} else {
		if dt == 0 {
			// '0' || BE(...)
			output.AppendBit(0) // Flag: compressed
		}

		// Determine extraction mask based on ct
		ct := comp.computeCtFlag(Vt, params.NewMaskFlag)

		if ct != 0 && Vt > 0 {
			// BE(It, (Xt OR Mt)) - extract bits where mask OR changes are set
			comp.workExtractMask.ORInto(comp.mask, Xt)
			BitExtract(output, input, comp.workExtractMask)
		} else {
			// BE(It, Mt) - extract only unpredictable bits
			BitExtract(output, input, comp.mask)
		}
	}

	// ================================================================
	// STEP 3: Update State for Next Cycle
	// ================================================================

	// Save current input and mask as previous for next iteration
	comp.prevInput.CopyFrom(input)
	comp.prevMask.CopyFrom(comp.mask)

	// Track new_mask_flag for ct calculation
	if params.NewMaskFlag {
		comp.newMaskFlagHistory[comp.flagHistoryIndex] = 1
	} else {
		comp.newMaskFlagHistory[comp.flagHistoryIndex] = 0
	}
	comp.flagHistoryIndex = (comp.flagHistoryIndex + 1) % MaxVtHistory

	// Advance time
	comp.t++

	// Advance history index (circular buffer)
	comp.historyIndex = (comp.historyIndex + 1) % MaxHistory

	return output.ToBytes(), nil
}

// computeRobustnessWindowInto computes Xt = OR of recent change vectors into dst.
func (comp *Compressor) computeRobustnessWindowInto(currentChange *BitVector, dst *BitVector) *BitVector {
	if comp.robustness == 0 || comp.t == 0 {
		// Xt = Dt (no reversal - RLE processes LSB to MSB directly)
		dst.CopyFrom(currentChange)
	} else {
		// Start with current change
		dst.CopyFrom(currentChange)

		// Determine how many historical changes to include
		numChanges := comp.t
		if comp.robustness < numChanges {
			numChanges = comp.robustness
		}

		// OR with historical changes (going backwards from current)
		for i := 1; i <= numChanges; i++ {
			// Calculate index of change from i iterations ago
			histIdx := (comp.historyIndex + MaxHistory - i) % MaxHistory
			// OR in place: dst = dst OR changeHistory[histIdx]
			for w := 0; w < dst.numWords; w++ {
				dst.data[w] |= comp.changeHistory[histIdx].data[w]
			}
		}
	}

	return dst
}

// leftShiftInto computes left shift of src into dst.
func leftShiftInto(dst, src *BitVector) {
	// Word-level left shift (big-endian: MSB in high bits of first word)
	var carry uint32
	for i := src.numWords - 1; i >= 0; i-- {
		word := src.data[i]
		dst.data[i] = ((word << 1) | carry) & 0xFFFFFFFF
		carry = (word >> 31) & 1
	}
}

// computeEffectiveRobustness computes Vt = Rt + Ct.
func (comp *Compressor) computeEffectiveRobustness(currentChange *BitVector) int {
	_ = currentChange // Unused - Ct is computed from history

	Rt := comp.robustness
	Vt := Rt

	// For t > Rt, compute Ct
	if comp.t > Rt {
		// Count backwards through history starting from Rt+1 positions back
		Ct := 0

		maxI := 15
		if comp.t < maxI {
			maxI = comp.t
		}

		for i := Rt + 1; i <= maxI; i++ {
			histIdx := (comp.historyIndex + MaxHistory - i) % MaxHistory
			if comp.changeHistory[histIdx].HammingWeight() > 0 {
				break // Found a change, stop counting
			}
			Ct++
			if Ct >= 15-Rt {
				break // Cap at maximum Ct value
			}
		}

		Vt = Rt + Ct
		if Vt > 15 {
			Vt = 15 // Cap at 15 (4 bits)
		}
	}

	return Vt
}

// computeCtFlag computes ct flag for multiple mask updates.
func (comp *Compressor) computeCtFlag(Vt int, currentNewMaskFlag bool) int {
	if Vt == 0 {
		return 0
	}

	// Count how many times new_mask_flag was set
	count := 0

	// Include current packet's flag
	if currentNewMaskFlag {
		count++
	}

	// Check history for Vt previous entries
	iterationsToCheck := Vt
	if comp.t < iterationsToCheck {
		iterationsToCheck = comp.t
	}

	for i := 0; i < iterationsToCheck; i++ {
		// Calculate history index going backwards from previous
		histIdx := (comp.flagHistoryIndex + MaxVtHistory - 1 - i) % MaxVtHistory
		if comp.newMaskFlagHistory[histIdx] != 0 {
			count++
		}
	}

	if count >= 2 {
		return 1
	}
	return 0
}

// hasPositiveUpdates checks for positive mask updates (et flag).
func hasPositiveUpdates(Xt, mask *BitVector) int {
	// et = 1 if any changed bits (in Xt) are predictable (mask bit = 0)
	for i := 0; i < Xt.length; i++ {
		bitChanged := Xt.GetBit(i)
		bitPredictable := mask.GetBit(i) == 0

		if bitChanged != 0 && bitPredictable {
			return 1 // Found a positive update
		}
	}
	return 0
}

// Internal helper functions that modify vectors in place

func updateBuildInternal(build, inputVec, prevInput, workChanges *BitVector, newMaskFlag bool, t int) {
	if t == 0 || newMaskFlag {
		build.Zero()
	} else {
		// Bt = (It XOR It-1) OR Bt-1
		workChanges.XORInto(inputVec, prevInput)
		build.ORInto(workChanges, build)
	}
}

func updateMaskInternal(mask, inputVec, prevInput, buildPrev, workChanges *BitVector, newMaskFlag bool) {
	// Calculate changes: It XOR It-1 (into workChanges)
	workChanges.XORInto(inputVec, prevInput)

	if newMaskFlag {
		// Mt = changes OR Bt-1
		mask.ORInto(workChanges, buildPrev)
	} else {
		// Mt = changes OR Mt-1 (in place)
		for w := 0; w < mask.numWords; w++ {
			mask.data[w] = workChanges.data[w] | mask.data[w]
		}
	}
}

func computeChangeInternal(change, mask, prevMask *BitVector, t int) {
	if t == 0 {
		// At t=0, D0 = M0 (assuming M-1 = 0)
		change.CopyFrom(mask)
	} else {
		// Dt = Mt XOR Mt-1
		change.XORInto(mask, prevMask)
	}
}

File: implementations/java/src/main/java/space/tanagra/pocketplus/Compressor.java
/*
 * Copyright (c) 2025 Tanagra Space
 * SPDX-License-Identifier: MIT
 */

package space.tanagra.pocketplus;

/**
 * POCKET+ compressor implementing CCSDS 124.0-B-1 Section 5.3.
 *
 * <p>Output packet format: oₜ = hₜ ∥ qₜ ∥ uₜ
 *
 * <ul>
 *   <li>hₜ = RLE(Xₜ) ∥ BIT₄(Vₜ) ∥ eₜ ∥ kₜ ∥ cₜ ∥ ḋₜ
 *   <li>qₜ = mask transmission (if needed)
 *   <li>uₜ = unpredictable bits or full input
 * </ul>
 */
public final class Compressor {

  private static final int MAX_HISTORY = 16;
  private static final int MAX_VT_HISTORY = 16;

  // Configuration
  private final int length;
  private final int robustness;
  private final BitVector initialMask;
  private final int ptLimit;
  private final int ftLimit;
  private final int rtLimit;

  // State
  private BitVector mask;
  private BitVector prevMask;
  private BitVector build;
  private BitVector prevInput;
  private final BitVector[] changeHistory;
  private final boolean[] flagHistory;
  private int historyIndex;
  private int flagHistoryIndex;
  private int t;

  // Countdown counters (match C reference)
  private int ptCounter;
  private int ftCounter;
  private int rtCounter;

  // Pre-allocated work buffers (avoid per-packet allocation)
  private final BitVector workPrevBuild;
  private final BitVector workChange;
  private final BitVector workXt;
  private final BitVector workShifted;
  private final BitVector workDiff;
  private final BitVector workExtractMask;

  /**
   * Creates a new Compressor.
   *
   * @param length packet length in bits
   * @param initialMask initial mask vector (null for all zeros)
   * @param robustness robustness parameter R (0-7)
   * @param ptLimit Pt parameter limit
   * @param ftLimit Ft parameter limit
   * @param rtLimit Rt parameter limit
   */
  public Compressor(
      int length, BitVector initialMask, int robustness, int ptLimit, int ftLimit, int rtLimit) {
    if (length <= 0 || length > PocketPlus.MAX_PACKET_LENGTH) {
      throw new IllegalArgumentException("Invalid packet length: " + length);
    }
    if (robustness < 0 || robustness > PocketPlus.MAX_ROBUSTNESS) {
      throw new IllegalArgumentException("Invalid robustness: " + robustness);
    }

    this.length = length;
    this.robustness = robustness;
    this.initialMask = initialMask != null ? new BitVector(initialMask) : new BitVector(length);
    this.ptLimit = ptLimit;
    this.ftLimit = ftLimit;
    this.rtLimit = rtLimit;

    // Initialize state
    this.mask = new BitVector(length);
    this.prevMask = new BitVector(length);
    this.build = new BitVector(length);
    this.prevInput = new BitVector(length);
    this.changeHistory = new BitVector[MAX_HISTORY];
    this.flagHistory = new boolean[MAX_VT_HISTORY];
    for (int i = 0; i < MAX_HISTORY; i++) {
      this.changeHistory[i] = new BitVector(length);
    }

    // Pre-allocate work buffers
    this.workPrevBuild = new BitVector(length);
    this.workChange = new BitVector(length);
    this.workXt = new BitVector(length);
    this.workShifted = new BitVector(length);
    this.workDiff = new BitVector(length);
    this.workExtractMask = new BitVector(length);

    reset();
  }

  /** Resets the compressor to initial state. */
  public void reset() {
    t = 0;
    historyIndex = 0;
    flagHistoryIndex = 0;

    // Reset mask to initial
    copyBitVector(mask, initialMask);
    prevMask.zero();
    build.zero();
    prevInput.zero();

    // Clear history
    for (int i = 0; i < MAX_HISTORY; i++) {
      changeHistory[i].zero();
    }
    for (int i = 0; i < MAX_VT_HISTORY; i++) {
      flagHistory[i] = false;
    }

    // Reset counters (match C reference: start at limit)
    ptCounter = ptLimit;
    ftCounter = ftLimit;
    rtCounter = rtLimit;
  }

  /**
   * Compresses a single packet using automatic parameter management.
   *
   * @param input the input packet as a BitVector
   * @return compressed output as a BitBuffer
   */
  public BitBuffer compressPacket(BitVector input) {
    // Compute parameters automatically
    boolean sendMaskFlag;
    boolean uncompressedFlag;
    boolean newMaskFlag;

    if (t == 0) {
      // First packet: fixed init values
      sendMaskFlag = true;
      uncompressedFlag = true;
      newMaskFlag = false;
    } else {
      // Check and update countdown counters (C reference style)
      if (ftCounter == 1) {
        sendMaskFlag = true;
        ftCounter = ftLimit;
      } else {
        ftCounter--;
        sendMaskFlag = false;
      }

      if (ptCounter == 1) {
        newMaskFlag = true;
        ptCounter = ptLimit;
      } else {
        ptCounter--;
        newMaskFlag = false;
      }

      if (rtCounter == 1) {
        uncompressedFlag = true;
        rtCounter = rtLimit;
      } else {
        rtCounter--;
        uncompressedFlag = false;
      }

      // CCSDS: first R+1 packets require ft=1, rt=1, pt=0
      if (t <= robustness) {
        sendMaskFlag = true;
        uncompressedFlag = true;
        newMaskFlag = false;
      }
    }

    return compressPacketWithParams(input, newMaskFlag, sendMaskFlag, uncompressedFlag);
  }

  /**
   * Compresses a packet with explicit parameters.
   *
   * @param input input packet
   * @param newMaskFlag whether to start new mask period
   * @param sendMaskFlag whether to send full mask
   * @param uncompressedFlag whether to send uncompressed
   * @return compressed output
   */
  public BitBuffer compressPacketWithParams(
      BitVector input, boolean newMaskFlag, boolean sendMaskFlag, boolean uncompressedFlag) {

    BitBuffer output = new BitBuffer();

    // Save previous mask and build before update (use pre-allocated buffers)
    prevMask.copyFrom(mask);
    workPrevBuild.copyFrom(build);

    // Update build vector (Equation 6) if t > 0
    if (t > 0) {
      MaskOperations.updateBuildInPlace(build, input, prevInput, newMaskFlag);
    }

    // Update mask vector (Equation 7) if t > 0
    if (t > 0) {
      MaskOperations.updateMaskInPlace(mask, input, prevInput, workPrevBuild, newMaskFlag);
    }

    // Compute change vector (Equation 8) into pre-allocated buffer
    MaskOperations.computeChangeInto(workChange, mask, prevMask, t);

    // Store change in history
    changeHistory[historyIndex].copyFrom(workChange);

    // === Encode output packet ===
    // oₜ = hₜ ∥ qₜ ∥ uₜ

    // Compute Xₜ (robustness window) into pre-allocated buffer
    computeRobustnessWindowInto(workXt, workChange);

    // Compute Vₜ (effective robustness)
    int vt = computeEffectiveRobustness();

    // Compute ḋₜ flag
    int dt = (!sendMaskFlag && !uncompressedFlag) ? 1 : 0;

    // === Component hₜ ===
    // hₜ = RLE(Xₜ) ∥ BIT₄(Vₜ) ∥ eₜ ∥ kₜ ∥ cₜ ∥ ḋₜ

    // 1. RLE(Xₜ)
    Encoder.rleEncode(output, workXt);

    // 2. BIT₄(Vₜ) - use bulk append
    output.appendBits(vt, 4);

    // 3. eₜ, kₜ, cₜ (only if Vt > 0 and there are changes)
    if (vt > 0 && workXt.hammingWeight() > 0) {
      // eₜ: positive updates exist?
      boolean et = hasPositiveUpdates(workXt, mask);
      output.appendBit(et ? 1 : 0);

      if (et) {
        // kₜ: extract inverted mask at change positions
        // Use workExtractMask as temp for inverted mask
        workExtractMask.copyFrom(mask);
        invertBitVector(workExtractMask);
        Encoder.bitExtractForward(output, workExtractMask, workXt);

        // cₜ: new_mask_flag was set 2+ times in last Vt+1 iterations
        int ct = computeCtFlag(vt, newMaskFlag);
        output.appendBit(ct);
      }
    }

    // 4. ḋₜ
    output.appendBit(dt);

    // === Component qₜ ===
    if (dt == 0) {
      if (sendMaskFlag) {
        output.appendBit(1); // Flag: mask follows
        // RLE(M XOR (M<<)) using pre-allocated buffers
        workShifted.copyFrom(mask);
        workShifted.leftShiftInPlace();
        workDiff.copyFrom(mask);
        workDiff.xorInPlace(workShifted);
        Encoder.rleEncode(output, workDiff);
      } else {
        output.appendBit(0); // Flag: no mask
      }
    }

    // === Component uₜ ===
    if (uncompressedFlag) {
      // '1' ∥ COUNT(F) ∥ Iₜ
      output.appendBit(1);
      Encoder.countEncode(output, length);
      output.appendBitVector(input, length);
    } else {
      if (dt == 0) {
        output.appendBit(0); // Flag: compressed
      }

      // Determine extraction mask using pre-allocated buffer
      int ct = computeCtFlag(vt, newMaskFlag);
      if (ct != 0 && vt > 0) {
        workExtractMask.copyFrom(mask);
        workExtractMask.orInPlace(workXt);
        Encoder.bitExtract(output, input, workExtractMask);
      } else {
        Encoder.bitExtract(output, input, mask);
      }
    }

    // === Update state ===
    copyBitVector(prevInput, input);

    // Track new_mask_flag
    flagHistory[flagHistoryIndex] = newMaskFlag;
    flagHistoryIndex = (flagHistoryIndex + 1) % MAX_VT_HISTORY;

    // Advance time and history
    t++;
    historyIndex = (historyIndex + 1) % MAX_HISTORY;

    return output;
  }

  private void computeRobustnessWindowInto(BitVector xt, BitVector currentChange) {
    // Start with current change
    xt.copyFrom(currentChange);

    if (robustness > 0 && t > 0) {
      // OR with historical changes using in-place operation
      int numChanges = Math.min(t, robustness);
      for (int i = 1; i <= numChanges; i++) {
        int histIdx = (historyIndex + MAX_HISTORY - i) % MAX_HISTORY;
        xt.orInPlace(changeHistory[histIdx]);
      }
    }
  }

  private void invertBitVector(BitVector bv) {
    // In-place bit inversion
    for (int i = 0; i < bv.numWords(); i++) {
      bv.setWord(i, ~bv.getWord(i));
    }
  }

  private int computeEffectiveRobustness() {
    // Vₜ = Rₜ + Cₜ
    int vt = robustness;

    if (t > robustness) {
      // Count consecutive iterations with no mask changes
      int ct = 0;
      for (int i = robustness + 1; i <= 15 && i <= t; i++) {
        int histIdx = (historyIndex + MAX_HISTORY - i) % MAX_HISTORY;
        if (changeHistory[histIdx].hammingWeight() > 0) {
          break;
        }
        ct++;
        if (ct >= 15 - robustness) {
          break;
        }
      }
      vt = robustness + ct;
      if (vt > 15) {
        vt = 15;
      }
    }

    return vt;
  }

  private boolean hasPositiveUpdates(BitVector xt, BitVector maskVec) {
    // eₜ = 1 if any changed bits have mask=0
    int numWords = Math.min(xt.numWords(), maskVec.numWords());
    for (int i = 0; i < numWords; i++) {
      if ((xt.getWord(i) & ~maskVec.getWord(i)) != 0) {
        return true;
      }
    }
    return false;
  }

  private int computeCtFlag(int vt, boolean currentNewMaskFlag) {
    // cₜ = 1 if new_mask_flag was set 2+ times in last Vt+1 iterations
    if (vt == 0) {
      return 0;
    }

    int count = currentNewMaskFlag ? 1 : 0;
    int itersToCheck = Math.min(vt, t);

    for (int i = 0; i < itersToCheck; i++) {
      int histIdx = (flagHistoryIndex + MAX_VT_HISTORY - 1 - i) % MAX_VT_HISTORY;
      if (flagHistory[histIdx]) {
        count++;
      }
    }

    return count >= 2 ? 1 : 0;
  }

  private void copyBitVector(BitVector dest, BitVector src) {
    dest.copyFrom(src);
  }

  /**
   * Returns the current time index.
   *
   * @return time index t
   */
  public int getTimeIndex() {
    return t;
  }
}
